python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -676,106 +670,65 @@ def _submit_runs_and_update_backfill_in_chunks(
-
-    # Initially, the only requested partitions are the partitions requested during the last
-    # backfill iteration
-    backfill_data_with_submitted_runs = (
-        asset_backfill_iteration_result.backfill_data.replace_requested_subset(submitted_partitions)
-    )
-
-    # Fetch backfill status
-    backfill = cast(PartitionBackfill, instance.get_backfill(backfill_id))
-    mid_iteration_cancel_requested = backfill.status != BulkActionStatus.REQUESTED
-    retryable_error_raised = False
+    run_requests: Sequence[RunRequest] = asset_backfill_iteration_result.run_requests
 
     # Iterate through runs to request, submitting runs in chunks.
     # In between each chunk, check that the backfill is still marked as 'requested',
     # to ensure that no more runs are requested if the backfill is marked as canceled/canceling.
+
+    num_submitted = 0
+
     for submit_run_request_chunk_result in submit_asset_runs_in_chunks(
         run_requests=run_requests,
-        reserved_run_ids=None,
+        reserved_run_ids=asset_backfill_iteration_result.reserved_run_ids,
         chunk_size=RUN_CHUNK_SIZE,
         instance=instance,
         workspace_process_context=workspace_process_context,
         asset_graph=asset_graph,
-        logger=logger,
         debug_crash_flags={},
-        backfill_id=backfill_id,
+        logger=logger,
+        run_tags={**run_tags, BACKFILL_ID_TAG: backfill_id},
     ):
-        if submit_run_request_chunk_result is None:
-            # allow the daemon to heartbeat
-            yield None
-            continue
-
-        retryable_error_raised = submit_run_request_chunk_result.retryable_error_raised
-
-        requested_partitions_in_chunk = _get_requested_asset_partitions_from_run_requests(
-            [rr for (rr, _) in submit_run_request_chunk_result.chunk_submitted_runs],
-            asset_graph,
-            instance_queryer,
-        )
-        submitted_partitions = submitted_partitions | AssetGraphSubset.from_asset_partition_set(
-            set(requested_partitions_in_chunk), asset_graph=asset_graph
-        )
+        if isinstance(submit_run_request_chunk_result, SubmitRunRequestChunkResult):
+            num_submitted += len(submit_run_request_chunk_result.chunk_submitted_runs)
+
+            # Indicates that a chunk has finished
+            # Refetch, in case the backfill was requested for cancellation in the meantime
+            backfill = cast(PartitionBackfill, instance.get_backfill(backfill_id))
+            if backfill.status != BulkActionStatus.REQUESTED:
+                # Move the run requests that we have not gotten to yet out of the requested subset
+                unsubmitted_run_ruquests = run_requests[num_submitted:]
+                unsubmitted_asset_partitions = _get_requested_asset_partitions_from_run_requests(
+                    unsubmitted_run_ruquests,
+                    asset_graph,
+                    instance_queryer,
+                )
```

Comment:
this is fallout from writing the asset backfill data before we actually submit the runs - if we determine that the backfill was moved into a canceling state, we have to move them back out or we'll hit an error case later that gets confused about there being requested partitions that don't have any runs, and the backfill will fail instead of being nicely canceled (I think there may be other edge cases with the code currently in master that lead to that state already, at least i have seen failed backfills that hit that while they were canceling)